File: bindings/go/tekken_test.go
package ztok

// Tekken (`tekken.json`) tokenizer tests for the Go binding.
//
// Loads the real mistral_nemo_tekken.json fixture (skipped when absent)
// and asserts ztok reproduces the canonical reference encodings, which
// were verified against mistral_common 1.8.6. Also checks that
// auto-detect routes Tekken vocabs to OpenTekken.

import (
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"testing"
)

// fileExists reports whether path names an existing file.
func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

// tekkenVocabPath returns the path to the real Tekken fixture under
// bench/vocabs/. bindings/go/ -> repo root -> bench/vocabs/...
func tekkenVocabPath() string {
	_, thisFile, _, _ := runtime.Caller(0)
	root := filepath.Dir(filepath.Dir(filepath.Dir(thisFile)))
	return filepath.Join(root, "bench", "vocabs", "mistral_nemo_tekken.json")
}

func openTekkenFixture(t testing.TB) *Pipeline {
	t.Helper()
	path := tekkenVocabPath()
	if !fileExists(path) {
		t.Skipf("Tekken vocab fixture not present at %s", path)
	}
	p, err := OpenTekken(path, nil)
	if err != nil {
		t.Fatalf("OpenTekken(%q): %v", path, err)
	}
	return p
}

// goldenTekken mirrors the cross-binding parity vectors: ids verified
// against mistral_common 1.8.6 on mistral_nemo_tekken.json.
var goldenTekken = []struct {
	text string
	want []uint32
}{
	{"Hello, world!", []uint32{22177, 1044, 4304, 1033}},
	{"The quick brown fox", []uint32{1784, 7586, 22980, 94137}},
	{" and the", []uint32{1321, 1278}},
}

func TestOpenTekkenGolden(t *testing.T) {
	p := openTekkenFixture(t)
	defer p.Close()

	for _, tc := range goldenTekken {
		ids, err := p.Encode(tc.text)
		if err != nil {
			t.Fatalf("encode %q: %v", tc.text, err)
		}
		if !reflect.DeepEqual(ids, tc.want) {
			t.Errorf("Encode(%q) = %v, want %v", tc.text, ids, tc.want)
		}
	}
}

func TestTekkenAutoDetect(t *testing.T) {
	path := tekkenVocabPath()
	if !fileExists(path) {
		t.Skipf("Tekken vocab fixture not present at %s", path)
	}
	if got := DetectFormat(path); got != FormatTekken {
		t.Fatalf("DetectFormat = %v, want FormatTekken", got)
	}
	// Open should dispatch to OpenTekken via auto-detect and match the
	// golden ids.
	p, err := Open(path)
	if err != nil {
		t.Fatalf("Open auto-detect: %v", err)
	}
	defer p.Close()

	ids, err := p.Encode("Hello, world!")
	if err != nil {
		t.Fatalf("encode after auto-detect Open: %v", err)
	}
	want := []uint32{22177, 1044, 4304, 1033}
	if !reflect.DeepEqual(ids, want) {
		t.Errorf("auto-detect Encode = %v, want %v", ids, want)
	}
}

File: bindings/nodejs/test/tekken.test.js
'use strict';

// Mistral Tekken tokenizer tests for the Node.js binding. Loads the real
// mistral_nemo_tekken.json fixture (skipped when absent) and checks ztok
// reproduces the golden ids verified against mistral_common 1.8.6.

const fs = require('fs');
const path = require('path');
const test = require('node:test');
const assert = require('node:assert/strict');

const ztok = require('..');

// bindings/nodejs/test/ -> repo root -> bench/vocabs/...
const VOCAB = path.resolve(
    __dirname, '..', '..', '..', 'bench', 'vocabs', 'mistral_nemo_tekken.json'
);

// Golden id sequences verified against mistral_common 1.8.6 on
// bench/vocabs/mistral_nemo_tekken.json.
const GOLDEN = [
    ['Hello, world!', [22177, 1044, 4304, 1033]],
    ['The quick brown fox', [1784, 7586, 22980, 94137]],
    [' and the', [1321, 1278]],
];

test('Tekken matches reference golden encodings', (t) => {
    if (!fs.existsSync(VOCAB)) {
        t.skip(`Tekken vocab fixture not present at ${VOCAB}`);
        return;
    }
    const pipe = ztok.Pipeline.fromTekken(VOCAB);
    try {
        for (const [text, want] of GOLDEN) {
            assert.deepEqual(Array.from(pipe.encode(text)), want, `encode mismatch for ${text}`);
        }
    } finally {
        pipe.close();
    }
});

test('Tekken auto-detects via detectFormat + fromPath', (t) => {
    if (!fs.existsSync(VOCAB)) {
        t.skip(`Tekken vocab fixture not present at ${VOCAB}`);
        return;
    }
    assert.equal(ztok.detectFormat(VOCAB), 'tekken');
    const pipe = ztok.Pipeline.fromPath(VOCAB);
    try {
        assert.deepEqual(Array.from(pipe.encode('Hello, world!')), [22177, 1044, 4304, 1033]);
    } finally {
        pipe.close();
    }
});
